broker/src/tunneldigger_broker/broker.py

```diff
@@ -77,6 +83,26 @@ def create_tunnel(self, broker, address, uuid, remote_tunnel_id, client_features
         else:
             tunnel_str = "(%s)" % uuid
 
+        # Rate limit creation of new tunnels with the same IP address.
+        # Runs broker.tunnel-connection-threshold hook if threshold exceeded:
+        if self.connection_rate_limit_per_ip_count > 0 and self.connection_rate_limit_per_ip_time > 0:
+            if address[0] not in self.last_tunnel_created_per_ip:
+                self.last_tunnel_created_per_ip[address[0]] = deque([now]) # Create deque collection if IP address does not exist in dict
+            if len(self.last_tunnel_created_per_ip[address[0]])-1 is self.connection_rate_limit_per_ip_count:
+                scale_delta = now - self.last_tunnel_created_per_ip[address[0]][0] # Delta of oldest timestamp in collection and now
+                if scale_delta <= self.connection_rate_limit_per_ip_time:
+                    logger.info("Tunnel %s connection threshold exceeded (%s in %s (%s) seconds). Running hook" % (tunnel_str, self.connection_rate_limit_per_ip_count, int(scale_delta), self.connection_rate_limit_per_ip_time))
```

Comment:
The `%s (%s)` here is not very clean.
Also, please use `"...".format()` instead of `%`-formatting.